### PR TITLE
Switch from register_full_backward_hooks to tensor hooks

### PR DESCRIPTION
**Pull request by vivekmig**
This switches usage of full backward hooks to instead apply forward hooks which then add tensor backward hooks, as suggested in #914 . We initially did not choose this approach since it may have limitations with backward hooks on modules with multiple tensors as inputs / outputs (each tensor must be called independently in the hook), but all current use-cases within Captum only require a single tensor input / output.

This change allows us to enable in-place modules as well as remove the limitation on neuron input attribution. DeepLift also no longer needs valid module checks, as these are no longer applicable with usage of tensor hooks.

### CODE REVIEW

File: captum/attr/_utils/lrp_rules.py [hunk 1]
```diff
@@ -53,6 +47,8 @@ def _backward_hook_input(grad):
                 self.relevance_input[device] = relevance.data
             else:
                 self.relevance_input[device].append(relevance.data)
+
+            grad.replace_out = relevance
```
Comment (NarineK) [inline captum/attr/_utils/lrp_rules.py:51]: I wonder why are we setting `replace_out` here. It looks like `_create_backward_hook_input` is called for supported linear layers and `backward_hook_activation` is for supported non-linear layers. Is this because we expect that the output of non-linear layer is the input of the linear layer ?

Maybe we can document the logic here ?

https://github.com/pytorch/captum/blob/460a170224a0b64b01af85f23e5005df44a4056e/captum/attr/_core/lrp.py#L320

**Review by NarineK — APPROVED** [review]
Thank you for working on this PR, @vivekmig! Looks much cleaner now, especially DeepLift.

Added couple questions and minor comments.

File: captum/_utils/common.py [hunk 2]
```diff
@@ -660,20 +660,52 @@ def _get_module_from_name(model: Module, layer_name: str) -> Any:
...
+
+        if isinstance(out, tuple):
+            assert (
+                len(out) == 1
+            ), "Backward hooks not supported for module with >1 output"
+            out[0].register_hook(output_tensor_hook)
+        else:
+            out.register_hook(output_tensor_hook)
+
+    def pre_hook(module, inp):
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/_utils/common.py:689]: Good point, will add type hints here

File: captum/_utils/common.py [hunk 3]
```diff
@@ -660,20 +660,52 @@ def _get_module_from_name(model: Module, layer_name: str) -> Any:
 
 def _register_backward_hook(
     module: Module, hook: Callable, attr_obj: Any
-) -> torch.utils.hooks.RemovableHandle:
+) -> List[torch.utils.hooks.RemovableHandle]:
     # Special case for supporting output attributions for neuron methods
     # This can be removed after deprecation of neuron output attributions
     # for NeuronDeepLift, NeuronDeconvolution, and NeuronGuidedBackprop
     # in v0.6.0
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/_utils/common.py:667]: Good point, will remove this comment

File: captum/_utils/common.py [hunk 4]
```diff
@@ -660,20 +660,52 @@ def _get_module_from_name(model: Module, layer_name: str) -> Any:
...
+
+        if isinstance(inp, tuple):
+            assert (
+                len(inp) == 1
+            ), "Backward hooks not supported for module with >1 input"
+            inp[0].register_hook(input_tensor_hook)
+            return inp[0].clone()
+        else:
+            inp.register_hook(input_tensor_hook)
+            return inp.clone()
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/_utils/common.py:725]: This clone is needed for handling inplace modules, otherwise the hook is not executed correctly.

File: captum/_utils/gradient.py [hunk 5]
```diff
@@ -37,6 +37,10 @@ def apply_gradient_requirements(
     ), "Inputs should be wrapped in a tuple prior to preparing for gradients"
     grad_required = []
     for index, input in enumerate(inputs):
+        if skip_non_tensor and not isinstance(input, torch.Tensor):
+            grad_required.append(None)
+            continue
+
         assert isinstance(input, torch.Tensor), "Given input is not a torch.Tensor"
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/_utils/gradient.py]: You are right, I think we can just remove this skip_non_tensor flag completely, I'm not sure why I thought it was necessary for TracIn. I've removed it now, thanks!

File: captum/attr/_utils/lrp_rules.py [hunk 6]
```diff
@@ -33,16 +33,10 @@ def forward_hook(self, module, inputs, outputs):
...
-        if (
-            isinstance(grad_input, tuple)
-            and isinstance(grad_output, tuple)
-            and len(grad_input) > len(grad_output)
-        ):
-            # Adds any additional elements of grad_input if applicable
-            # This occurs when registering a backward hook on nn.Dropout
-            # modules, which has an additional element of None in
-            # grad_input
-            return grad_output + grad_input[len(grad_output) :]
```
[the remark this answers is not in the document]
Comment (vivekmig) [reply captum/attr/_utils/lrp_rules.py:45]: We don't currently, I will look into adding them in a separate diff. This shouldn't be a problem with the new approach, since that was a quirk of the module hook; we now set the hooks from the forward hook which should be fine.

File: captum/attr/_utils/lrp_rules.py (hunk 1 above)
Comment (vivekmig) [reply captum/attr/_utils/lrp_rules.py:51]: Yes, the issue is that there are 2 tensor backward hooks on the same tensor, but the output doesn't get propagated as input to the next hook automatically, so need this workaround. Will add some comments to explain it better, thanks!
